pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,226 @@
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/robfig/cron"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+//isCloudKubernetesServiceIndexer
+const (
+	isCKSIndexer = "clusters.management.cattle.io/is-cks"
+)
+
+var (
+	clusterUpstreamRefresher *clusterRefreshController
+)
+
+// for other cloud drivers, please edit HERE
+type clusterConfig struct {
+	eksConfig *eksv1.EKSClusterConfigSpec
+}
+
+func init() {
+	// possible settings controller, which references refresh
+	// cron job, will run prior to StartClusterUpstreamCronJob.
+	// This ensure the CronJob will not be nil
+	clusterUpstreamRefresher = &clusterRefreshController{
+		refreshCronJob: cron.New(),
+	}
+}
+
+type clusterRefreshController struct {
+	refreshCronJob *cron.Cron
+	secretsCache   wranglerv1.SecretCache
+	clusterClient  v3.ClusterClient
+	clusterCache   v3.ClusterCache
+}
+
+func StartClusterUpstreamCronJob(wContext *wrangler.Context) {
+	clusterUpstreamRefresher.secretsCache = wContext.Core.Secret().Cache()
+	clusterUpstreamRefresher.clusterClient = wContext.Mgmt.Cluster()
+	clusterUpstreamRefresher.clusterCache = wContext.Mgmt.Cluster().Cache()
+
+	clusterUpstreamRefresher.clusterCache.AddIndexer(isCKSIndexer, func(obj *apimgmtv3.Cluster) ([]string, error) {
+		// for other cloud drivers, please edit HERE
+		// if obj.Spec.EKSConfig == nil && obj.Spec.GKEConfig == nil && ...
+		if obj.Spec.EKSConfig == nil {
+			return []string{}, nil
+		}
+		return []string{"true"}, nil
+	})
+
+	schedule, err := cron.ParseStandard(settings.ClusterUpstreamRefreshCron.Get())
+	if err != nil {
+		logrus.Errorf("Error parsing upstream cluster refresh cron. Upstream state will not be refreshed: %v", err)
+		return
+	}
+	clusterUpstreamRefresher.refreshCronJob.Schedule(schedule, cron.FuncJob(clusterUpstreamRefresher.refreshAllUpstreamStates))
+	clusterUpstreamRefresher.refreshCronJob.Start()
+}
+
+func (c *clusterRefreshController) refreshAllUpstreamStates() {
+	logrus.Debugf("Refreshing clusters' upstream states")
+	clusters, err := c.clusterCache.GetByIndex(isCKSIndexer, "true")
+	if err != nil {
+		logrus.Error("error trying to refresh clusters' upstream states")
+		return
+	}
+
+	for _, cluster := range clusters {
+		if _, err := c.refreshClusterUpstreamSpec(cluster); err != nil {
+			logrus.Errorf("error refreshing cluster [%s] upstream state", cluster.Name)
+		}
+	}
+}
+
+func (c *clusterRefreshController) refreshClusterUpstreamSpec(cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	// for other cloud drivers, please edit HERE
+	// if obj.Spec.EKSConfig == nil && obj.Spec.GKEConfig == nil && ...
+	if cluster.Spec.EKSConfig == nil {
+		return cluster, nil
+	}
```

Comment:
I don't think this is necessary if the index key is more descriptive. Just trying to descrease the number of places that need to be modified when we add new hosted providers.

Reply:
Please look and review on new